### PR TITLE
htlcswitch: set sourceRef when resolving packets through interceptor

### PR DESCRIPTION
**Pull request by champo**
Fixes https://github.com/lightningnetwork/lnd/issues/5115

The problem was several layers removed from the fix. The link decides to
clean up a `fwdPkg` only if it's completed, otherwise it renotifies the
HTLCs. A package is only set to complete if it's `addAck` and
`settleFail` filters are full. For forwarded HTLCs, the `addAck` was
never being set so it would never be considered complete under this
criteria.

`addAck` is set for an HTLC when signing the next commitment TX in the
`LightningChannel`. The path for this is:
* `LightningChannel#SettleHtlc` adds the HTLC to `localUpdates`
* `LightningChannel#SignNextCommitment` builds the `ackAddRef` for all
updates with `SourceRef != nil`.
* `LightningChannel#SignNextCommitment` then passes the list of
`ackAddRef` to `OpenChannel#AppendRemoteCommitChain` to persist the new
acks in the filter

Since `SourceRef` was nil for interceptor packages, `SignNextCommitment`
ignored it and the ack was never persisted.

### CODE REVIEW

**Comment by champo** [comment]
Does this warrant an itest? Not sure what the criteria is.

**Comment by Roasbeef** [comment]
Ping @cfromknecht

**Comment by carlaKC** [comment]
> Does this warrant an itest? Not sure what the criteria is.

I think it would be good to update the existing itest to replicate the behaviour we're fixing here so that we can assert that it's been fixed? 

> I also have 0 clue as to how to fix already broken forward packages that doesn't mean closing all channels 

So the issue here is that we now have these intercepted forward packages that have a nil source ref, and are being re-forwarded despite no longer being active in the commit? Just making sure I'm following.

**Review by carlaKC — COMMENTED** [review]
Change looks good, an addition to the itest would be great! From the original issue, I think that we can just restart `bob` in our three hop network to reproduce this?

**Comment by champo** [comment]
> > Does this warrant an itest? Not sure what the criteria is.
> 
> 
> 
> I think it would be good to update the existing itest to replicate the behaviour we're fixing here so that we can assert that it's been fixed? 
> 

OK. will look into it!

> 
> > I also have 0 clue as to how to fix already broken forward packages that doesn't mean closing all channels 
> 
> 
> 
> So the issue here is that we now have these intercepted forward packages that have a nil source ref, and are being re-forwarded despite no longer being active in the commit? Just making sure I'm following. 
> 
> 

That's exactly it.

**Comment by champo** [comment]
@carlaKC added an extra check after the existing interceptor test. Run it a few times with & without the patch and the output was as desired.

**Comment by Crypt-iQ** [comment]
Haven't reviewed the test, but this seems like the right fix.  I don't think we should patch the forwarding package code to handle old buggy interceptable-switch forwarding packages though.  Adding more logic / complexity is undesirable IMO.  Though it might be possible for us to give *you* a patch to run to clean up your forwarding packages, rather than it living in the repo for an indefinite amount of time.

**Comment by champo** [comment]
> Haven't reviewed the test, but this seems like the right fix. I don't think we should patch the forwarding package code to handle old buggy interceptable-switch forwarding packages though. Adding more logic / complexity is undesirable IMO. Though it might be possible for us to give _you_ a patch to run to clean up your forwarding packages, rather than it living in the repo for an indefinite amount of time.

Sounds reasonable to me. I'd write the patch but I honestly have no idea what the invariants are in that part of the code and feels like something easily messed up. We have several nodes that should have HTLCs in this state that we could test the patches on.

**Comment by Roasbeef** [comment]
> So the issue here is that we now have these intercepted forward packages that have a nil source ref, and are being re-forwarded despite no longer being active in the commit? Just making sure I'm following.

@champo so are y'all seeing _every intercepted HTLC ever_ being re-forwarded _each time_ lnd restarts?

File: lntest/itest/lnd_forward_interceptor_test.go [hunk 1]
```diff
@@ -224,6 +225,45 @@ func testForwardInterceptor(net *lntest.NetworkHarness, t *harnessTest) {
...
+
+	err = restartAlice()
+	require.NoError(t.t, err, "failed to restart alice")
+
+	go func() {
+		request, err := interceptor.Recv()
+		if err != nil {
+			// If it is  just the error result of the context cancellation
+			// the we exit silently.
+			status, ok := status.FromError(err)
```
Comment (Roasbeef) [inline lntest/itest/lnd_forward_interceptor_test.go:249]: 👍

File: lntest/itest/lnd_forward_interceptor_test.go [hunk 2]
```diff
@@ -224,6 +225,45 @@ func testForwardInterceptor(net *lntest.NetworkHarness, t *harnessTest) {
 	// that tests the payment final status for the held payment.
 	cancelInterceptor()
 	wg.Wait()
+
+	// Verify that we don't get notified about already completed HTLCs
```
Comment (Roasbeef) [inline lntest/itest/lnd_forward_interceptor_test.go:229]: Test addition is solid, only a few style nits here to ensure lines are wrapped to 80 columns (line 262, etc). 

Confirmed that w/o the prior commit, this test fails as those intercepted packets are being replayed.

**Review by Roasbeef — APPROVED** [review]
LGTM ☄️

Just a few style nits, would like to package this with our upcoming 0.13.1 release!

**Comment by champo** [comment]
> @champo so are y'all seeing _every intercepted HTLC ever_ being re-forwarded _each time_ lnd restarts?

Yeah. It's a bit better & a bit worse in practice. If a channel is closed all those are forgotten, so it's not every ever. Buuut every time a peer restarts we get all the intercepted HTLCs from that channel reforwarded too.

Will address the comments tomorrow morning.

**Pull request merged**